### ISSUE TITLE
AuthorizationCode flow doesn't return email scope/claim.

### ISSUE THREAD

**Issue by myaesubi** [IdentityServer/IdentityServer3#3383]
Hi,
This a question...

I've used the AspNetIdentity identity sample for my idp server. The client that connects to the idp uses '_**AuthorizationCode**_' flow.
As I see the identityserver3 log, the handshake happens correctly and successful, however the created JWT does not have email claim and the client errors out with 

> "email claim retrieved from access token is null or empty. Verify your identity provider provides the email claim with the given the OpenID scope settings. "

I've setup the client in IDP to get pretty much all AllowedScopes.

The following is the information from the log:
w3wp.exe Information: 0 : 2016-11-23 04:02:56.406 +00:00 [Information] "Authorize request validation success"
 "{
  \"ClientId\": \"client.sp\",
  \"ClientName\": \"client\",
  \"RedirectUri\": \"https://client.com/auth/openIdLogin\",
  \"AllowedRedirectUris\": [
    \"https://client.com/auth/openIdLogin\"
  ],
  \"SubjectId\": \"1\",
  \"ResponseType\": \"code\",
  \"ResponseMode\": \"query\",
  \"Flow\": \"AuthorizationCode\",
  \"RequestedScopes\": \"openid email profile\",
  \"State\": \"path=/&XSRF-TOKEN=mooU5a1UUH9LluL08FvsG2wyDdln9NK4\",
  \"Nonce\": \"1TTdGVIggZc\",
  \"SessionId\": \"8984d19affb5fad10e57be6760e4acda\",
  \"Raw\": {
    \"response_type\": \"code\",
    \"client_id\": \"client.sp\",
    \"redirect_uri\": \"https://client.com/auth/openIdLogin\",
    \"scope\": \"openid email profile\",
    \"state\": \"path=/&XSRF-TOKEN=mooU5a1UUH9LluL08FvsG2wyDdln9NK4\",
    \"nonce\": \"1TTdGVIggZc\"
  }
}"
w3wp.exe Information: 0 : 2016-11-23 04:02:56.754 +00:00 [Information] Creating Authorization Code Flow response.
2016-11-23 04:02:56.784 +00:00 [Debug] Adding client "client.sp" to client list cookie for subject "1"
w3wp.exe Information: 0 : 2016-11-23 04:02:56.815 +00:00 [Information] End authorize request
w3wp.exe Information: 0 : 2016-11-23 04:02:56.815 +00:00 [Information] Redirecting to: https://client.com/auth/openIdLogin
w3wp.exe Information: 0 : 2016-11-23 04:02:57.213 +00:00 [Information] Start token request
2016-11-23 04:02:57.244 +00:00 [Debug] Start client validation
2016-11-23 04:02:57.276 +00:00 [Debug] Start parsing Basic Authentication secret
2016-11-23 04:02:57.276 +00:00 [Debug] Parser found secret: "BasicAuthenticationSecretParser"
w3wp.exe Information: 0 : 2016-11-23 04:02:57.276 +00:00 [Information] Secret id found: "client.sp"
2016-11-23 04:02:57.307 +00:00 [Debug] Secret validator success: "HashedSharedSecretValidator"
w3wp.exe Information: 0 : 2016-11-23 04:02:57.324 +00:00 [Information] Client validation success
w3wp.exe Information: 0 : 2016-11-23 04:02:57.354 +00:00 [Information] Start token request validation
w3wp.exe Information: 0 : 2016-11-23 04:02:57.416 +00:00 [Information] Start validation of authorization code token request
w3wp.exe Information: 0 : 2016-11-23 04:02:57.463 +00:00 [Information] Validation of authorization code token request success
w3wp.exe Information: 0 : 2016-11-23 04:02:57.496 +00:00 [Information] Token request validation success
 {
  "ClientId": "client.sp",
  "ClientName": "client",
  "GrantType": "authorization_code",
  "AuthorizationCode": "13073fe4d16bd778f97f80d64130c95a",
  "Raw": {
    "code": "13073fe4d16bd778f97f80d64130c95a",
    "redirect_uri": "https://client.com/auth/openIdLogin",
    "grant_type": "authorization_code"
  }
}

Appreciate it a lot to learn what I am missing in the implementation of Idp server.

Thanks,
Mohammad

**Comment by leastprivilege** [comment]
Claims are returned in `IUserService.GetProfileDataAsync` - that's where I would put my breakpoint.

**Comment by myaesubi** [comment]
Hello again,

Thanks for the reply. I had the NuGet package for "IdentityServer3.AspNetIdentity" which comes both as .CS and NuGet package.

I setup the logging for more detailed logging and here is out:
w3wp.exe Information: 0 : 2016-11-30 02:52:42.812 +00:00 [Information] {
  "Category": "TokenService",
  "Name": "Access token issued",
  "EventType": "Information",
  "Id": 2000,
  "Details": {
    "TokenType": "Jwt",
    "SubjectId": "1",
    "ClientId": "client.sp",
    "Scopes": [
      "openid",
      "email",
      "profile"
    ],
    "Lifetime": 60,
    "Claims": {
      "client_id": "client.sp",
      "scope": [
        "openid",
        "email",
        "profile"
      ],
      "sub": "1",
      "auth_time": 1480474361,
      "idp": "idsrv",
      "amr": "password"
    }
  },
  "Context": {
    "ActivityId": "00000000-0000-0000-7c50-0080000000ec",
    "TimeStamp": "2016-11-30T02:52:42.8128689+00:00",
    "ProcessId": 7776,
    "MachineName": "RD000D3A125CCE",
    "RemoteIpAddress": "104.41.144.99"
  }
}

I see that 'email' claim is not there. I feel I have a **blindspot** here that following the AspNetIdentity sample code still doesn't send the email claim; 

Thanks again,
Mohammad

**Comment by myaesubi** [comment]
Thanks @leastprivilege 

As I debugged the AspNetIdenty sample, with breakpoint on the `IUserService.GetProfileDataAsync`, when flow is 'AuthroizationCode' the break point doesn't go there.
I tested using the HybrdFlow and was able to include the 'email' claim redefined to be always included.

The challenge is that the client product that is using my idp server only support '**AuthorizationCode**' flow and I'm looking to be able inject the '**email**' claim; 
now I'm reading the identityserver3 source code looking for a spot to be able to inject 'email' claim.

appreciate any hint

Regards,
Mohammad
